system/security/plugins/jwtSecurity/jwtCache.cpp

```diff
@@ -0,0 +1,392 @@
+    std::string     s2 = std::regex_replace(s1, std::regex("\\[[0-9].+?\\]\\*?"), "1");
+    std::string     s3 = std::regex_replace(s2, std::regex("\\[![A-Za-z].+?\\]\\*?"), "1");
+    std::string     s4 = std::regex_replace(s3, std::regex("\\[[A-Za-z].+?\\]\\*?"), "a");
+    std::string     s5 = std::regex_replace(s4, std::regex("[*?]"), "a");
+
+    return s5;
+}
+
+std::string JWTUserInfo::_convertPathname(const std::string& logicalPath) const
+{
+    std::string newPath(std::regex_replace(logicalPath, std::regex("::"), "/"));
+
+    if (!newPath.empty())
+    {
+        if (newPath.back() == '/')
+            newPath += "*";
+        else if (newPath.back() != '*')
+            newPath += "/*";
+    }
+
+    return newPath;
+}
+
+JWTUserInfo& JWTUserInfo::_addScopePerm(PermissionMap& scopePermMap, const std::string& scope, SecAccessFlags accessFlag, SecAccessFlags defaultFlag, bool asDeny)
+{
+    std::string                 rewrittenScope(_convertPathname(scope));
+    PermissionMap::iterator     foundIter = scopePermMap.find(rewrittenScope);
+
+    if (foundIter != scopePermMap.end())
+    {
+        // Exact match on scope; apply access to existing bitmask
+        if (asDeny)
+            foundIter->second = static_cast<SecAccessFlags>(foundIter->second & ~accessFlag);
+        else
+            foundIter->second = static_cast<SecAccessFlags>(foundIter->second | accessFlag);
+    }
+    else
+    {
+        // Find most-applicable scope access value and apply new access to that bitmask
+        SecAccessFlags  existingFlags = _matchScopePerm(scopePermMap, scope, defaultFlag);
+
+        if (asDeny)
+            scopePermMap[rewrittenScope] = static_cast<SecAccessFlags>(existingFlags & ~accessFlag);
+        else
+            scopePermMap[rewrittenScope] = static_cast<SecAccessFlags>(existingFlags | accessFlag);
+    }
+
+    return *this;
+}
+
+JWTUserInfo& JWTUserInfo::_addScopePerm(PermissionMap& scopePermMap, const std::string& scope, const std::string& accessFlagName, SecAccessFlags defaultFlag, bool asDeny)
+{
+    return _addScopePerm(scopePermMap, scope, getSecAccessFlagValue(accessFlagName.c_str()), defaultFlag, asDeny);
+}
+
+SecAccessFlags JWTUserInfo::_matchScopePerm(const PermissionMap& scopePermMap, const std::string& scope, SecAccessFlags defaultFlag) const
+{
+    // Ideally, this function should be traversing a tree of scope permissions, with each
+    // scope level a tree node; that would be much faster and would also avoid requiring
+    // ordering multiple patterns prior to calling _addScopePerm()
+    std::string         scopeAsExample(_convertPathname(_globToExample(scope)));
+    SecAccessFlags      accessValue = defaultFlag;
+    int                 longestScopeLen = 0;
+
+    for (PermissionMap::const_iterator x = scopePermMap.begin(); x != scopePermMap.end(); x++)
+    {
+        if (::fnmatch(x->first.c_str(), scopeAsExample.c_str(), (FNM_CASEFOLD | FNM_LEADING_DIR)) == 0)
```

Comment:
Is this available on all platforms, or do we disable on windows?  Is there a better, more portable way to do the same, e,g. using std::regex, which could be precalulated in the scope permissions.

Reply:
The original design was to leverage file globbing rather than regex for patterns because they are more intuitive (primarily because globbing is not greedy).  Windows supposedly does have a similar function -- PathMatchSpecA() -- but I am not at all sure it _really_ functions the same as fnmatch() and I lack a Windows development environment to test.

My suspicion is that if someone is running a cluster on Windows then they are more likely to use Active Directory than JWT as an authentication/authorization scheme.  Therefore, I would argue that this plugin should be disabled for Windows.

That said, I can be convinced otherwise.  Do we have a hand-rolled pattern matcher in jlib I can use instead?  If so, that may  be an even better route.